### PR TITLE
Added option --max-sleep-interval to support random sleep between downloads. Fixes #9930

### LINKED ISSUE #9930: Allow Random Values for --sleep-interval
The `--sleep-interval` option is great. I think that this is the main reason why we can run downloads from heavily monitored sites.

Is there a way to modify this option so that it can accept random values? For example, instead of setting it to `--sleep-interval 300` to set it to sleep for 300 seconds or 5 minutes between each download, make it like so:

`--sleep-interval 120-300`

so that the sleep interval would be set between 120 and 300 seconds.

The rationale is to simulate more realistic browsing behavior.

### PR DESCRIPTION
**Pull request by singh-pratyush96**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] New extractor
- [x] New feature

---

### Added option `--max-sleep-interval` to support random sleep between downloads.

The new option allows user to perform sleep between consecutive downloads in the interval `[a, b]` where `0 <= a <= b` and are set by the options `--sleep-interval` and `--max-sleep-interval` respectively.

#### Example usage :
```bash
youtube-dl https://www.youtube.com/playlist?list=PLsPUh22kYmNDnezZ15MkicBGVDm2ypthG --sleep-interval 10 --max-sleep-interval 20
```

### CODE REVIEW

File: youtube_dl/downloader/common.py [hunk 1]
```diff
@@ -342,8 +343,10 @@ def download(self, filename, info_dict):
             })
             return True
 
-        sleep_interval = self.params.get('sleep_interval')
-        if sleep_interval:
+        sleep_lower_bound = self.params.get('sleep_interval')
+        if sleep_lower_bound:
+            sleep_upper_bound = self.params.get('max_sleep_interval') if self.params.get('max_sleep_interval') else sleep_lower_bound
```
Comment (dstftw) [inline youtube_dl/downloader/common.py]: Just `sleep_upper_bound = self.params.get('max_sleep_interval', sleep_lower_bound)`.

File: youtube_dl/options.py [hunk 2]
```diff
@@ -502,6 +502,11 @@ def _scrub_eq(o):
         '--sleep-interval', metavar='SECONDS',
```
Comment (dstftw) [inline youtube_dl/options.py]: `--min-sleep-interval` should be added for symmetry.

File: youtube_dl/options.py [hunk 3]
```diff
@@ -502,6 +502,11 @@ def _scrub_eq(o):
...
         dest='sleep_interval', type=float,
         help='Number of seconds to sleep before each download.')
```
Comment (dstftw) [inline youtube_dl/options.py]: Description should be clarified for the case when using with `--max-sleep-interval`.

**Comment by singh-pratyush96** [comment]
I have made the changes.

PS: Please excuse me for the silly mistakes I make being a newbie.

**Comment by dstftw** [comment]
Could you squash commits?

File: youtube_dl/YoutubeDL.py [hunk 4]
```diff
@@ -249,7 +249,11 @@ class YoutubeDL(object):
     source_address:    (Experimental) Client-side IP address to bind to.
     call_home:         Boolean, true iff we are allowed to contact the
                        youtube-dl servers for debugging.
-    sleep_interval:    Number of seconds to sleep before each download.
+    sleep_interval:    Minimum number of seconds to sleep before each download.
+                       Sleep will be for a random interval if --max-sleep-interval is also passed.
+    max_sleep_interval:Max number of seconds to sleep before each download.
+                       Sleep will be for a random interval if passed along with --min-sleep-interval
+                       or --sleep-interval, otherwise ignored.
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: These are programmatic options not command line. CLI should not even be mentioned here.

**Comment by singh-pratyush96** [comment]
No, I couldn't squash. Rebasing merges all the changes into single commit but the intermediate commits still exist.

But as I saw while merging some PR, Github provides an option to squash all the commits in the PR before merging.

File: youtube_dl/YoutubeDL.py (hunk 4 above)
Comment (singh-pratyush96) [inline youtube_dl/YoutubeDL.py]: Does 'these' refer to `--max-sleep-interval` and `--min-sleep-interval`?

Where are they supposed to be mentioned in the case (`options.py`)?

**Comment by dstftw** [comment]
We try to avoid GitHub merge due to merge commits it introduces. Anyway if GitHub can do that obviously this can be done by hand and you just doint it wrong.

File: youtube_dl/YoutubeDL.py (hunk 4 above)
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: They are already mentioned in `options.py`.

Comment (singh-pratyush96) [inline youtube_dl/YoutubeDL.py]: In the case, shall I remove `max_sleep_interval:Max number...` comment from here?

Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: You should remove mention of any command line switches but not programmatic options as I've already said.

**Comment by dstftw** [comment]
Cherry-picked in 065bc354894f1d35592529455d9eb685470124b9, thanks.

**Pull request closed without merging**